spec/features/registration/sign_up_as_an_organisation_with_enhanced_2fa_spec.rb

```diff
@@ -0,0 +1,243 @@
+require "support/notifications_service"
+require "support/confirmation_use_case"
+
+describe "Sign up as an organisation", type: :feature do
+  let(:name) { "Sally" }
+
+  before do
+    allow(Rails.application.config).to receive(:enable_enhanced_2fa_experience).and_return true
+    Rails.application.config.s3_aws_config = {
+      stub_responses: {
+        get_object: { body: SIGNUP_WHITELIST_PREFIX_MATCHER + '(gov\.uk)$' },
+      },
+    }
+  end
+
+  after { Warden.test_reset! }
+
+  include_context "when sending a confirmation email"
+  include_context "when using the notifications service"
+
+  context "with a valid email" do
+    let(:email) { "newuser@gov.uk" }
+
+    before { sign_up_for_account(email: email) }
+
+    it "instructs the user to check their confirmation email" do
+      expect(page).to have_content(
+        "A confirmation email has been sent to your email address.",
+      )
+    end
+  end
+
+  context "with correct data" do
+    before do
+      sign_up_for_account(email: email)
+      update_user_details(name: name)
+    end
+
+    context "with a gov.uk email" do
+      let(:email) { "someone@gov.uk" }
+
+      it "signs me in" do
+        expect(page).to have_content "Sign out"
+      end
+
+      it "creates a confirmed membership joining the user to the org" do
+        user = User.find_by(email: email)
+        organisation = Organisation.find_by(name: "Gov Org 1")
+        expect(user.membership_for(organisation)).to be_confirmed
+      end
+    end
+
+    context "with a email from a subdomain of gov.uk" do
+      let(:email) { "someone@other.gov.uk" }
+
+      it "signs me in" do
+        expect(page).to have_content "Sign out"
+      end
+    end
+
+    context "with a non-gov email" do
+      let(:email) { "someone@google.com" }
+
+      it "tells me my email is not valid" do
+        expect(page).to have_content("Email address must be from a government or a public sector domain. If you're having trouble signing up, contact us.")
+      end
+
+      it "provides a link to a support form" do
+        visit new_user_registration_path
+        click_on "Sign up"
+        click_on "contact us"
+        expect(page).to have_content("Setting up GovWifi in your organisation support")
+      end
+    end
+
+    context "with a blank email" do
+      let(:email) { "" }
+
+      it "tells me my email is not valid" do
+        expect(page).to have_content("Email address must be from a government or a public sector domain. If you're having trouble signing up, contact us.")
+      end
+    end
+
+    context "without a name" do
+      let(:name) { "" }
+      let(:email) { "someone@other.gov.uk" }
```

Comment:
all of this looks like unit tests for the User model to me. Given how expensive feature tests are, we definitely don't want to enforce all of our validation at the system level.

Reply:
That's a good point, but this was taken from an existing test in order to make it work specifically with the new 2FA journeys. Happy to consider this outside this PR.